fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java

```diff
@@ -193,18 +199,70 @@ public void afterHistory(FHIRPersistenceEvent event) throws FHIRPersistenceInter
+                    Resource resource = entry.getResource();
+                    if (resource instanceof Provenance) {
+                        // skip enforcement of provenance resources, but keep track of their targets
+                        provenanceTargets.add(((Provenance) resource).getTarget());
+                    } else {
+                        enforce(resource, getPatientIdFromToken(jwt), Permission.READ, getScopesFromToken(jwt));
+                        relativeResourcePaths.add(resource.getClass().getSimpleName() + "/" + resource.getId());
+                    }
                 }
             }
+
+            enforceProvenance(provenanceTargets, relativeResourcePaths);
         } else {
             throw new IllegalStateException("Expected event resource of type Bundle but found " +
                     event.getFhirResource().getClass().getSimpleName());
         }
     }
 
+    /**
+     * If the search result contains one or more Provenance resources,
+     * then reject the request unless each Provenance targets at least one resource in the response bundle
+     *
+     * @throws IllegalStateException if the baseUrl cannot be computed from the request context
+     * @throws FHIRPersistenceInterceptorException if the list of provenanceTargets contains
+     */
+    private void enforceProvenance(List<List<Reference>> provenanceTargets, Set<String> relativeResourcePaths)
+            throws FHIRPersistenceInterceptorException {
+        if (provenanceTargets.isEmpty()) return;
+
+        String baseUrl;
+        try {
+            baseUrl = ReferenceUtil.getServiceBaseUrl();
+        } catch (FHIRSearchException e) {
+            throw new IllegalStateException("Unexpected error while computing the service baseUrl for ");
+        }
+
+        for (List<Reference> targets : provenanceTargets) {
+            boolean allow = false;
+            for (Reference r : targets) {
+                ReferenceValue referenceValue = ReferenceUtil.createReferenceValueFrom(r, baseUrl);
+                if (ReferenceValue.ReferenceType.LITERAL_RELATIVE == referenceValue.getType()) {
+                    if (relativeResourcePaths.contains(referenceValue.getTargetResourceType() + "/" + referenceValue.getValue())) {
+                        allow = true;
+                        break;
+                    }
+                } else if (log.isLoggable(Level.FINE)){
+                    log.fine("Skipping target '" + referenceValue.getValue() + "' of type '" + referenceValue.getType() +
+                            "' during enforcement of Provenance access.");
+                }
+            }
+
+            if (!allow) {
+                throw new FHIRPersistenceInterceptorException(REQUEST_NOT_PERMITTED)
+                        .withIssue(FHIRUtil.buildOperationOutcomeIssue(REQUEST_NOT_PERMITTED, IssueType.FORBIDDEN));
+            }
+        }
+    }
+
```

Comment:
I think its right.  I could have passed a List of Provenances to this function, but for better or worse I just pulled the target from each Provenance resource a priori.  Because the `if (!allow)` is within the for loop, it will make the decision on a single Provenance.target (List<Reference>) at a time and deny access if any one doesn't meet the criteria.

Reply:
I also agree that maybe we don't even need this level of checking; because searching for Provenance resources directly will still get scoped to the Patient compartment, the only way to get a Provenance resource in the response would be to _revinclude it which would mean its target is there by definition

Reply:
I'd pull it out (with comment about why we don't need to check it), but OK by me if you want to leave it in.

Reply:
OK, I pulled it out with my latest commit.  I also added support for enforcing direct access to Provenance resources; please have another look when you get a chance.